# DAT ASS: Algebraic Modeling

In [4]:
%reload_ext sql

In [5]:
%sql postgresql://appdev@data/appdev

'Connected: appdev@appdev'

## Question 1.

In [10]:
%sql SELECT * FROM chinook.track WHERE genreid = 18 union SELECT * FROM chinook.track WHERE genreid = 20;

39 rows affected.


[(3248, 'Take the Celestra', 253, 3, 20, None, 2927677, 512381289, Decimal('1.99')),
 (3243, 'Murder On the Rising Star', 253, 3, 20, None, 2935894, 551759986, Decimal('1.99')),
 (3238, 'The Living Legend, Pt. 2', 253, 3, 20, None, 2923298, 515632754, Decimal('1.99')),
 (3230, 'Lost Planet of the Gods, Pt. 2', 253, 3, 20, None, 2914664, 534343985, Decimal('1.99')),
 (3249, 'The Hand of God', 253, 3, 20, None, 2924007, 536583079, Decimal('1.99')),
 (3226, 'Battlestar Galactica, Pt. 1', 253, 3, 20, None, 2952702, 541359437, Decimal('1.99')),
 (3242, 'The Man With Nine Lives', 253, 3, 20, None, 2956998, 577829804, Decimal('1.99')),
 (2837, 'Crossroads, Pt. 1', 227, 3, 20, None, 2622622, 486233524, Decimal('1.99')),
 (2827, 'Unfinished Business', 227, 3, 18, None, 2622038, 528499160, Decimal('1.99')),
 (2836, 'The Son Also Rises', 227, 3, 18, None, 2621830, 499258498, Decimal('1.99')),
 (2831, 'Taking a Break from All Your Worries', 227, 3, 18, None, 2624207, 492700163, Decimal('1.99')),
 (3228, 'Battlestar Galactica, Pt. 3', 253, 3, 20, None, 2927802, 554509033, Decimal('1.99')),
 (3227, 'Battlestar Galactica, Pt. 2', 253, 3, 20, None, 2956081, 521387924, Decimal('1.99')),
 (3239, 'Fire In Space', 253, 3, 20, None, 2926593, 536784757, Decimal('1.99')),
 (2838, 'Crossroads, Pt. 2', 227, 3, 20, None, 2869953, 497335706, Decimal('1.99')),
 (2834, 'Dirty Hands', 227, 3, 18, None, 2627961, 537648614, Decimal('1.99')),
 (2825, 'A Measure of Salvation', 227, 3, 18, None, 2563938, 489715554, Decimal('1.99')),
 (3244, 'Greetings from Earth, Pt. 1', 253, 3, 20, None, 2960293, 536824558, Decimal('1.99')),
 (3234, 'The Gun On Ice Planet Zero, Pt. 2', 253, 3, 20, None, 2924341, 546542281, Decimal('1.99')),
 (3241, 'War of the Gods, Pt. 2', 253, 3, 20, None, 2923381, 487899692, Decimal('1.99')),
 (3245, 'Greetings from Earth, Pt. 2', 253, 3, 20, None, 2903778, 527842860, Decimal('1.99')),
 (3235, 'The Magnificent Warriors', 253, 3, 20, None, 2924716, 570152232, Decimal('1.99')),
 (3231, 'The Lost Warrior', 253, 3, 20, None, 2920045, 558872190, Decimal('1.99')),
 (3247, 'Experiment In Terra', 253, 3, 20, None, 2923548, 547982556, Decimal('1.99')),
 (3246, "Baltar's Escape", 253, 3, 20, None, 2922088, 525564224, Decimal('1.99')),
 (2833, 'A Day In the Life', 227, 3, 18, None, 2620245, 462818231, Decimal('1.99')),
 (2829, 'The Eye of Jupiter', 227, 3, 18, None, 2618750, 517909587, Decimal('1.99')),
 (2828, 'The Passage', 227, 3, 18, None, 2623875, 490375760, Decimal('1.99')),
 (3233, 'The Gun On Ice Planet Zero, Pt. 1', 253, 3, 20, None, 2907615, 540980196, Decimal('1.99')),
 (3232, 'The Long Patrol', 253, 3, 20, None, 2925008, 513122217, Decimal('1.99')),
 (2819, 'Battlestar Galactica: The Story So Far', 226, 3, 18, None, 2622250, 490750393, Decimal('1.99')),
 (3229, 'Lost Planet of the Gods, Pt. 1', 253, 3, 20, None, 2922547, 537812711, Decimal('1.99')),
 (2830, 'Rapture', 227, 3, 18, None, 2624541, 508406153, Decimal('1.99')),
 (2826, 'Hero', 227, 3, 18, None, 2713755, 506896959, Decimal('1.99')),
 (2835, 'Maelstrom', 227, 3, 18, None, 2622372, 514154275, Decimal('1.99')),
 (2832, 'The Woman King', 227, 3, 18, None, 2626376, 552893447, Decimal('1.99')),
 (3236, 'The Young Lords', 253, 3, 20, None, 2863571, 587051735, Decimal('1.99')),
 (3240, 'War of the Gods, Pt. 1', 253, 3, 20, None, 2922630, 505761343, Decimal('1.99')),
 (3237, 'The Living Legend, Pt. 1', 253, 3, 20, None, 2924507, 503641007, Decimal('1.99'))]

In [26]:
%sql SELECT * FROM chinook.genre WHERE genreid = 18;

1 rows affected.


[(18, 'Science Fiction')]

In [25]:
%sql SELECT * FROM chinook.genre WHERE genreid = 20;

1 rows affected.


[(20, 'Sci Fi & Fantasy')]

The results seems to be scifi related soundtracks.

## Question 2.

In [24]:
%sql SELECT * FROM chinook.invoice WHERE total > 5 intersect SELECT * FROM chinook.invoice WHERE total < 10;

115 rows affected.


[(396, 18, datetime.datetime(2013, 10, 8, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', Decimal('8.91')),
 (290, 32, datetime.datetime(2012, 6, 27, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '696 Osborne Street', 'Winnipeg', 'MB', 'Canada', 'R3L 2B9', Decimal('5.94')),
 (185, 52, datetime.datetime(2011, 3, 20, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '202 Hoxton Street', 'London', None, 'United Kingdom', 'N1 5LH', Decimal('5.94')),
 (53, 44, datetime.datetime(2009, 8, 11, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Porthaninkatu 9', 'Helsinki', None, 'Finland', '00530', Decimal('8.91')),
 (10, 46, datetime.datetime(2009, 2, 3, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '3 Chatham Street', 'Dublin', 'Dublin', 'Ireland', None, Decimal('5.94')),
 (172, 41, datetime.datetime(2011, 1, 20, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '11, Place Bellecour', 'Lyon', None, 'France', '69002', Decimal('8.91')),
 (52, 38, datetime.datetime(2009, 8, 8, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Barbarossastraße 19', 'Berlin', None, 'Germany', '10779', Decimal('5.94')),
 (205, 44, datetime.datetime(2011, 6, 20, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Porthaninkatu 9', 'Helsinki', None, 'Finland', '00530', Decimal('7.96')),
 (94, 30, datetime.datetime(2010, 2, 10, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '230 Elgin Street', 'Ottawa', 'ON', 'Canada', 'K2P 1L7', Decimal('5.94')),
 (220, 6, datetime.datetime(2011, 8, 22, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', Decimal('5.94')),
 (11, 52, datetime.datetime(2009, 2, 6, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '202 Hoxton Street', 'London', None, 'United Kingdom', 'N1 5LH', Decimal('8.91')),
 (46, 6, datetime.datetime(2009, 7, 11, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', Decimal('8.91')),
 (332, 24, datetime.datetime(2012, 12, 30, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', Decimal('5.94')),
 (283, 53, datetime.datetime(2012, 5, 27, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '113 Lupus St', 'London', None, 'United Kingdom', 'SW1V 3EN', Decimal('5.94')),
 (39, 27, datetime.datetime(2009, 6, 10, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', Decimal('8.91')),
 (318, 7, datetime.datetime(2012, 10, 29, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Rotenturmstraße 4, 1010 Innere Stadt', 'Vienne', None, 'Austria', '1010', Decimal('5.94')),
 (347, 47, datetime.datetime(2013, 3, 5, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Via Degli Scipioni, 43', 'Rome', 'RM', 'Italy', '00192', Decimal('8.91')),
 (213, 27, datetime.datetime(2011, 7, 22, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', Decimal('5.94')),
 (192, 31, datetime.datetime(2011, 4, 20, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), '194A Chain Lake Drive', 'Halifax', 'NS', 'Canada', 'B3S 1C5', Decimal('5.94')),
 (199, 10, datetime.datetime(2011, 5, 21, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', Decimal('5.94')),
 (367, 37, datetime.datetime(2013, 6, 3, 0, 0, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None)), 'Berger Straße 10', 'Frankfurt', None, 'Germany', '60316', Decimal('5.94')),
 (234, 23, datetime.datetime(2011, 10, 23, 0, 0, tzinfo=psycopg2.tz.

The resulting invoices have similarities in their total value, fx 8.91 and 5.94.

## Question 3.

In [29]:
%sql SELECT * FROM chinook.customer WHERE country = 'USA' except SELECT * FROM chinook.customer WHERE email LIKE '%yahoo.com';

11 rows affected.


[(21, 'Kathy', 'Chase', None, '801 W 4th Street', 'Reno', 'NV', 'USA', '89503', '+1 (775) 223-7665', None, 'kachase@hotmail.com', 5),
 (22, 'Heather', 'Leacock', None, '120 S Orange Ave', 'Orlando', 'FL', 'USA', '32801', '+1 (407) 999-7788', None, 'hleacock@gmail.com', 4),
 (16, 'Frank', 'Harris', 'Google Inc.', '1600 Amphitheatre Parkway', 'Mountain View', 'CA', 'USA', '94043-1351', '+1 (650) 253-0000', '+1 (650) 253-0000', 'fharris@google.com', 4),
 (28, 'Julia', 'Barnett', None, '302 S 700 E', 'Salt Lake City', 'UT', 'USA', '84102', '+1 (801) 531-7272', None, 'jubarnett@gmail.com', 5),
 (19, 'Tim', 'Goyer', 'Apple Inc.', '1 Infinite Loop', 'Cupertino', 'CA', 'USA', '95014', '+1 (408) 996-1010', '+1 (408) 996-1011', 'tgoyer@apple.com', 3),
 (18, 'Michelle', 'Brooks', None, '627 Broadway', 'New York', 'NY', 'USA', '10012-2612', '+1 (212) 221-3546', '+1 (212) 221-4679', 'michelleb@aol.com', 3),
 (17, 'Jack', 'Smith', 'Microsoft Corporation', '1 Microsoft Way', 'Redmond', 'WA', 'USA', '98052-8300', '+1 (425) 882-8080', '+1 (425) 882-8081', 'jacksmith@microsoft.com', 5),
 (26, 'Richard', 'Cunningham', None, '2211 W Berry Street', 'Fort Worth', 'TX', 'USA', '76110', '+1 (817) 924-7272', None, 'ricunningham@hotmail.com', 4),
 (20, 'Dan', 'Miller', None, '541 Del Medio Avenue', 'Mountain View', 'CA', 'USA', '94040-111', '+1 (650) 644-3358', None, 'dmiller@comcast.com', 4),
 (27, 'Patrick', 'Gray', None, '1033 N Park Ave', 'Tucson', 'AZ', 'USA', '85719', '+1 (520) 622-4200', None, 'patrick.gray@aol.com', 4),
 (24, 'Frank', 'Ralston', None, '162 E Superior Street', 'Chicago', 'IL', 'USA', '60611', '+1 (312) 332-3232', None, 'fralston@gmail.com', 3)]

Surpisingly only 11 USA customers without a yahoo mail.

## Question 4.

In [50]:
%sql SELECT DISTINCT title FROM chinook.track, chinook.album WHERE composer LIKE '%Mozart%' union SELECT DISTINCT title FROM chinook.track, chinook.album WHERE composer LIKE '%Bach%'

347 rows affected.


[('The Ultimate Relexation Album',),
 ('Bach: Goldberg Variations',),
 ('Virtual XI',),
 ('Instant Karma: The Amnesty International Campaign to Save Darfur',),
 ('Judas 0: B-Sides and Rarities',),
 ('Weill: The Seven Deadly Sins',),
 ('Bongo Fury',),
 ('Powerslave',),
 ("Up An' Atom",),
 ('Original Soundtracks 1',),
 ('Arquivo II',),
 ('Vs.',),
 ('In Your Honor [Disc 2]',),
 ('Come Taste The Band',),
 ('King For A Day Fool For A Lifetime',),
 ('St. Anger',),
 ('Van Halen',),
 ('BBC Sessions [Disc 1] [Live]',),
 ('No More Tears (Remastered)',),
 ('Revelations',),
 ('My Generation - The Very Best Of The Who',),
 ('Pure Cult: The Best Of The Cult (For Rockers, Ravers, Lovers & Sinners) [UK]',),
 ('Stormbringer',),
 ('The Real Thing',),
 ('Aquaman',),
 ('The Song Remains The Same (Disc 1)',),
 ('Hot Rocks, 1964-1971 (Disc 1)',),
 ('Jagged Little Pill',),
 ('Outbreak',),
 ('Serie Sem Limite (Disc 1)',),
 ('Ao Vivo [IMPORT]',),
 ('Audioslave',),
 ('20th Century Masters - The Millennium Collection: The Best of Scorpions',),
 ('Minas',),
 ('J.S. Bach: Chaconne, Suite in E Minor, Partita in E Major & Prelude, Fugue and Allegro',),
 ('Slaves And Masters',),
 ('Diary of a Madman (Remastered)',),
 ('...And Justice For All',),
 ('In Through The Out Door',),
 ('The Best Of 1980-1990',),
 ('Bach: Toccata & Fugue in D Minor',),
 ('Motley Crue Greatest Hits',),
 ('Pop',),
 ('Respighi:Pines of Rome',),
 ('Sozinho Remix Ao Vivo',),
 ('House of Pain',),
 ('Allegri: Miserere',),
 ("All That You Can't Leave Behind",),
 ('Let There Be Rock',),
 ('Serie Sem Limite (Disc 2)',),
 ('Sir Neville Marriner: A Celebration',),
 ('Bach: The Brandenburg Concertos',),
 ('War',),
 ('Adams, John: The Chairman Dances',),
 ('Restless and Wild',),
 ('Killers',),
 ("Liszt - 12 Études D'Execution Transcendante",),
 ('Da Lama Ao Caos',),
 ('No Prayer For The Dying',),
 ('Unplugged',),
 ('Grieg: Peer Gynt Suites & Sibelius: Pelléas et Mélisande',),
 ('Chill: Brazil (Disc 1)',),
 ('Volume Dois',),
 ('UB40 The Best Of - Volume Two [UK]',),
 ('Supernatural',),
 ('Szymanowski: Piano Works, Vol. 1',),
 ('Wagner: Favourite Overtures',),
 ('Puccini: Madama Butterfly - Highlights',),
 ('Bizet: Carmen Highlights',),
 ('Raul Seixas',),
 ('Prokofiev: Symphony No.5 & Stravinksy: Le Sacre Du Printemps',),
 ("Great Performances - Barber's Adagio and Other Romantic Favorites for Strings",),
 ('Coda',),
 ('Berlioz: Symphonie Fantastique',),
 ('The Battle Rages On',),
 ('Un-Led-Ed',),
 ('Bach: The Cello Suites',),
 ('By The Way',),
 ('Rock In Rio [CD2]',),
 ('Armada: Music from the Courts of England and Spain',),
 ('Afrociberdelia',),
 ('Na Pista',),
 ('Djavan Ao Vivo - Vol. 1',),
 ('Bach: Orchestral Suites Nos. 1 - 4',),
 ('The Office, Season 1',),
 ('A Real Dead One',),
 ('Blizzard of Ozz',),
 ('Vozes do MPB',),
 ('Voodoo Lounge',),
 ('In Your Honor [Disc 1]',),
 ('Warner 25 Anos',),
 ('Physical Graffiti [Disc 2]',),
 ('The Song Remains The Same (Disc 2)',),
 ('Greatest Hits',),
 ('Battlestar Galactica (Classic), Season 1',),
 ('How To Dismantle An Atomic Bomb',),
 ('Nevermind',),
 ('Blood Sugar Sex Magik',),
 ('Vivaldi: The Four Seasons',),
 ('Tangents',),
 ('Santana Live',),
 ('Master Of Puppets',),
 ('Fear Of The Dark',),
 ('MK III The Final Concerts [Disc 1]',),
 ('Fireball',),
 ('Sex Machine',),
 ('Black Album',),
 ('The X Factor',),
 ('Acústico MTV [Live]',),
 ('Purcell: The Fairy Queen',),
 ('Great Opera Choruses',),
 ('Blue Moods',),
 ('Out Of Time',),
 ('Carmina Burana',),
 ('Physical Graffiti [Disc 1]',),
 ('Live After Death',),
 ('Into The Light',),
 ('Greatest Hits II',),
 ('South American Getaway',),
 ('Live At Donington 1992 (Disc 1)',),
 ('Cake: B-Sides and Rarities',),
 ('Strauss: Waltzes',),
 ('Minha História',),
 ('Handel: Music for the Royal Fireworks (Original Version 1749)',),
 ('The Number of The Beast',),
 ('Achtung Baby',),
 ('Cidade Negra - Hits',),
 ('Greatest Kiss',),
 ('Quanta Gente Veio Ver (Live)',),
 ('Misplaced Childhood',),
 ('Sambas De Enredo 2001',),

In [58]:
%sql SELECT DISTINCT * FROM chinook.track, chinook.album WHERE title = 'Judas 0: B-Sides and Rarities' AND (composer LIKE '%Bach' OR composer LIKE '%Mozart%')

12 rows affected.


[(3407, 'Concerto for 2 Violins in D Minor, BWV 1043: I. Vivace', 276, 2, 24, 'Johann Sebastian Bach', 193722, 3192890, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3408, 'Aria Mit 30 Veränderungen, BWV 988 "Goldberg Variations": Aria', 277, 2, 24, 'Johann Sebastian Bach', 120463, 2081895, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3409, 'Suite for Solo Cello No. 1 in G Major, BWV 1007: I. Prélude', 278, 2, 24, 'Johann Sebastian Bach', 143288, 2315495, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3412, '"Eine Kleine Nachtmusik" Serenade In G, K. 525: I. Allegro', 281, 2, 24, 'Wolfgang Amadeus Mozart', 348971, 5760129, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3413, 'Concerto for Clarinet in A Major, K. 622: II. Adagio', 282, 2, 24, 'Wolfgang Amadeus Mozart', 394482, 6474980, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3430, 'Toccata and Fugue in D Minor, BWV 565: I. Toccata', 297, 2, 24, 'Johann Sebastian Bach', 153901, 2649938, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3433, 'Concerto No.2 in F Major, BWV1047, I. Allegro', 300, 2, 24, 'Johann Sebastian Bach', 307244, 5064553, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3451, 'Die Zauberflöte, K.620: "Der Hölle Rache Kocht in Meinem Herze"', 317, 2, 25, 'Wolfgang Amadeus Mozart', 174813, 2861468, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3454, 'Symphony No. 41 in C Major, K. 551, "Jupiter": IV. Molto allegro', 320, 2, 24, 'Wolfgang Amadeus Mozart', 362933, 6173269, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3482, 'Suite No. 3 in D, BWV 1068: III. Gavotte I & II', 327, 2, 24, 'Johann Sebastian Bach', 225933, 3847164, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3490, 'Partita in E Major, BWV 1006A: I. Prelude', 335, 2, 24, 'Johann Sebastian Bach', 285673, 4744929, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131),
 (3502, 'Quintet for Horn, Violin, 2 Violas, and Cello in E Flat Major, K. 407/386c: III. Allegro', 346, 2, 24, 'Wolfgang Amadeus Mozart', 221331, 3665114, Decimal('0.99'), 201, 'Judas 0: B-Sides and Rarities', 131)]

The results showed many albums, containing tracks that were composed by Mozart or Bach.